File: game_of_life_c/src/game.c
#ifndef GAME_C
#define GAME_C

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "raylib.h"

const int N = 256;
const int M = N - 1;

const int CELL = 4;
const int SCREEN = N * CELL;

const uint8_t ALIVE = 1;
const uint8_t DEAD = 0;

typedef struct Game {
    uint8_t *curr;
    uint8_t *old;
} Game;

void init(Game *game) {
    game->curr = (uint8_t*)calloc(N * N, sizeof(uint8_t));
    game->old =(uint8_t*)calloc(N * N, sizeof(uint8_t));
    if (!game->curr || !game->old) {
        fprintf(stderr, "Out of memory");
        exit(0);
    }
}

void deinit(Game *game) {
    free(game->old);
    free(game->curr);
}

int idx(int row, int col) {
    return row * N + col;
}

void swap_buffers(Game *game) {
    uint8_t *temp = game->curr;
    game->curr = game->old;
    game->old = temp;
}

void randomize(Game *game) {
    srand(99706354);
    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < N; ++c) {
            game->curr[idx(r, c)] = (uint8_t)rand() & ALIVE;
        }
    }
}

void draw_cell(int row, int col) {
    DrawRectangle(col * CELL, row * CELL, CELL, CELL, RED);
}

uint8_t update_cell(Game *game, int row, int col) {
    const int n = row > 0 ? row - 1 : M;
    const int s = row < M ? row + 1 : 0;
    const int w = col > 0 ? col - 1 : M;
    const int e = col < M ? col + 1 : 0;

    uint8_t sum = 0;

    sum += game->old[idx(n, w)];
    sum += game->old[idx(n, col)];
    sum += game->old[idx(n, e)];

    sum += game->old[idx(row, w)];
    sum += game->old[idx(row, e)];

    sum += game->old[idx(s, w)];
    sum += game->old[idx(s, col)];
    sum += game->old[idx(s, e)];

    const uint8_t old = game->old[idx(row, col)];
    const uint8_t curr = (old == ALIVE && sum == 2) || sum == 3;

    return curr;
}

void draw_cells(Game *game) {
    swap_buffers(game);
    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < N; ++c) {
            uint8_t cell = update_cell(game, r, c);
            game->curr[idx(r, c)] = cell;
            if (cell == ALIVE) {
                draw_cell(r, c);
            }
        }
    }
}

#endif

File: game_of_life_zig/src/game.zig
const std = @import("std");

const ray = @cImport({
    @cInclude("raylib.h");
});

pub const N: usize = 256;
pub const M: usize = N - 1;

pub const CELL: c_int = 4;
pub const SCREEN: c_int = N * CELL;

pub const ALIVE: u8 = 1;
pub const DEAD: u8 = 0;

pub const Game = struct {
    allocator: std.mem.Allocator,
    rand: std.Random,
    curr: []u8,
    old: []u8,

    pub fn init(allocator: std.mem.Allocator, rand: std.Random) !Game {
        const curr = try allocator.alloc(u8, N * N);
        errdefer allocator.free(curr);

        const old = try allocator.alloc(u8, N * N);
        errdefer allocator.free(old);

        return .{
            .curr = curr,
            .old = old,
            .allocator = allocator,
            .rand = rand,
        };
    }

    pub fn deinit(self: Game) void {
        self.allocator.free(self.old);
        self.allocator.free(self.curr);
    }

    pub fn randomize(self: *Game) void {
        for (0..N) |row| {
            for (0..N) |col| {
                self.curr[index(row, col)] = self.rand.uintLessThan(u8, 2);
            }
        }
    }

    pub fn draw_cells(self: *Game) void {
        self.swap_buffers();

        for (0..N) |row| {
            for (0..N) |col| {
                const cell = self.update_cell(row, col);
                self.curr[index(row, col)] = cell;
                if (cell == ALIVE) {
                    draw_cell(row, col);
                }
            }
        }
    }

    fn update_cell(self: Game, row: usize, col: usize) u8 {
        const n: usize = if (row > 0) row - 1 else M;
        const s: usize = if (row < M) row + 1 else 0;
        const w: usize = if (col > 0) col - 1 else M;
        const e: usize = if (col < M) col + 1 else 0;

        var sum: u8 = 0;

        sum += self.old[index(n, w)];
        sum += self.old[index(n, col)];
        sum += self.old[index(n, e)];

        sum += self.old[index(row, w)];
        sum += self.old[index(row, e)];

        sum += self.old[index(s, w)];
        sum += self.old[index(s, col)];
        sum += self.old[index(s, e)];

        const old = self.old[index(row, col)];
        const curr: u8 = @intFromBool((old == ALIVE and sum == 2) or sum == 3);

        return curr;
    }

    fn draw_cell(row: usize, col: usize) void {
        const r: c_int = @intCast(row);
        const c: c_int = @intCast(col);
        ray.DrawRectangle(
            c * CELL,
            r * CELL,
            CELL,
            CELL,
            ray.GREEN,
        );
    }

    inline fn swap_buffers(self: *Game) void {
        const tmp: []u8 = self.old;
        self.old = self.curr;
        self.curr = tmp;
    }

    inline fn index(row: usize, col: usize) usize {
        return row * N + col;
    }
};
